private/bufpkg/bufimage/bufimageutil/image_filter.go

```diff
@@ -266,26 +264,31 @@ func (b *sourcePathsBuilder) remapDependencies(
 	dependencyChanges := make([]int32, len(dependencies))
 	for _, importPath := range dependencies {
 		path := append(dependencyPath, indexFrom)
-		if importsRequired != nil && importsRequired.index(importPath) != -1 {
+		if _, ok := importsRequired[importPath]; ok {
 			dependencyChanges[indexFrom] = indexTo
 			if indexTo != indexFrom {
 				sourcePathsRemap.markMoved(path, indexTo)
 			}
 			newDependencies = append(newDependencies, importPath)
 			indexTo++
 			// delete them as we go, so we know which ones weren't in the list
-			importsRequired.delete(importPath)
+			delete(importsRequired, importPath)
 		} else {
 			sourcePathsRemap.markDeleted(path)
 			dependencyChanges[indexFrom] = -1
 		}
 		indexFrom++
 	}
-	if importsRequired != nil {
-		newDependencies = append(newDependencies, importsRequired.keys()...)
+	// Add imports picked up via a public import. The filtered files do not use public imports.
+	if publicImportCount := len(importsRequired); publicImportCount > 0 {
+		for importPath := range importsRequired {
+			newDependencies = append(newDependencies, importPath)
+		}
+		// Sort the public imports to ensure the output is deterministic.
+		sort.Strings(newDependencies[len(newDependencies)-publicImportCount:])
```

Comment:
This ordering isn't consistent with the current ordering... which means that we could be creating a breaking behaviour here, right?

Reply:
The current ordering is not deterministic. The issue https://github.com/bufbuild/buf/issues/3725 was re-raised as the unused imports (which should be just public imports) were being added in an undefined order. The issue brought up would still affect any file that has 2 or more public imports (which should be hopefully no-one), but this aims to solve that edge case.